File: guided-demo/graphs-2/src/dfs.js
/**
 * Edge class
 */
class Edge {
	constructor(destination, weight=1) {
		this.destination = destination;
		this.weight = weight;
	}
}

/**
 * Vertex class
 */
class Vertex {
	constructor(value='vertex') {
		this.value = value;
		this.edges = [];
	}
}

/**
 * Graph class
 */
class Graph {
	constructor() {
		this.vertexes = [];
	}

	dfs(start) {
		function visit(v) {
			v.color = gray;

			for (e of v.edges) {
				const neighbor = e.destination;
				if (neighbor.color === 'white') {
					neighbor.parent = v;
					visit(neighbor);
				}
			}

			v.color = 'black';
		}

		for (v of this.vertexes) {
			v.color = 'white';
			v.parent = null;
		}

		for (v of this.vertexes) {
			if (v.color === 'white') {
				visit(v);
			}
		}
	}
}

/**
 * Helper function to add bidirectional edges
 */
function addEdge(v0, v1) {
	v0.edges.push(new Edge(v1));
	v1.edges.push(new Edge(v0));
}

/**
 * Main
 */

// Build graph
const graph = new Graph();
const vertA = new Vertex('A');
const vertB = new Vertex('B');
const vertC = new Vertex('C');
const vertD = new Vertex('D');
const vertE = new Vertex('E');
const vertF = new Vertex('F');
const vertG = new Vertex('G');
const vertH = new Vertex('H');

addEdge(vertA, vertB);
addEdge(vertB, vertD);
addEdge(vertA, vertC);
addEdge(vertC, vertD);
addEdge(vertC, vertF);
addEdge(vertG, vertF);
addEdge(vertE, vertF);
addEdge(vertH, vertF);
addEdge(vertH, vertE);

graph.vertexes.push(vertA);
graph.vertexes.push(vertB);
graph.vertexes.push(vertC);
graph.vertexes.push(vertD);
graph.vertexes.push(vertE);
graph.vertexes.push(vertF);
graph.vertexes.push(vertG);
graph.vertexes.push(vertH);

File: guided-demo/graphs-2/src/dfs.py
#!/usr/bin/python

"""
Represent a graph with depth-first-search, and run a demo with a randomly
generated graph containing a given number of vertices and edges (default 8).
"""

from sys import argv


class Vertex:
    """Represent a vertex with label and set of edges to other vertices."""
    # Just holds data, but is class and not namedtuple/dict so it is hashable
    # pylint: disable=too-few-public-methods
    def __init__(self, label):
        self.label = label
        self.edges = set()

    def __str__(self):
        return "Vertex " + str(self.label)


class Graph:
    """Represent a graph and enable depth-first search."""
    def __init__(self):
        self.vertices = set()

    def add_edge(self, vertex_a, vertex_b):
        """Add a *bidirectional* edge between two vertices."""
        if vertex_a not in self.vertices or vertex_b not in self.vertices:
            raise Exception('Vertices to connect not in graph!')
        vertex_a.edges.add(vertex_b)
        vertex_b.edges.add(vertex_a)

    def dfs(self, start, target=None):
        """Perform depth-first search from a start, to optional target."""
        if start not in self.vertices:
            raise Exception('Start vertex not in graph!')

        visit_stack = [start]
        visited = set()
        while visit_stack:
            visiting = visit_stack.pop()
            visited.add(visiting)
            visiting.color = 'gray'
            if target and visiting == target:
                print('Found target:', visiting)
                return
            visit_stack.extend(visiting.edges - visited)

        if target:
            print('Target not found!')


def main(num_vertices=8, num_edges=8):
    """Build demo graph and execute DFS."""
    graph = Graph()
    for num in range(num_vertices):
        graph.vertices.add(Vertex(label=str(num)))

    # Add some random edges
    from random import sample
    for _ in range(num_edges):
        vertices = sample(graph.vertices, 2)
        graph.add_edge(vertices[0], vertices[1])

    # DFS with random start/target
    start, target = sample(graph.vertices, 2)
    graph.dfs(start, target)


if __name__ == '__main__':
    if len(argv) == 3:
        NUM_VERTICES = argv[1]
        NUM_EDGES = argv[2]
        main(NUM_VERTICES, NUM_EDGES)
    else:
        main()  # accept defaults
